File: packages/runtime/src/bridge.mjs
export const BRIDGE_MAGIC = 0x0D000721;
export const BRIDGE_VERSION = 1;
export const BridgeOpcode = Object.freeze({ createNode: 1, deleteNode: 2, insertChild: 3, removeChild: 4, setProp: 5, setModifier: 6, setText: 7 });
const opcodeName = Object.fromEntries(Object.entries(BridgeOpcode).map(([k, v]) => [v, k]));

export function encodeBridgeBatch(ops) {
  const strings = []; const stringIndex = new Map(); const words = [];
  const intern = (value) => { const text = String(value ?? ""); if (!stringIndex.has(text)) { stringIndex.set(text, strings.length); strings.push(text); } return stringIndex.get(text); };
  for (const op of ops) {
    const code = BridgeOpcode[op.op]; if (!code) throw new Error(`Unknown bridge op ${op.op}`); words.push(code);
    switch (op.op) {
      case "createNode": words.push(op.id, intern(op.nodeType)); break;
      case "deleteNode": words.push(op.id); break;
      case "insertChild": words.push(op.parent, op.child, op.index); break;
      case "removeChild": words.push(op.parent, op.child); break;
      case "setProp": words.push(op.id, intern(op.key), intern(encodeValue(op.value))); break;
      case "setModifier": words.push(op.id, intern(encodeValue(op.modifier))); break;
      case "setText": words.push(op.id, intern(op.text)); break;
    }
  }
  const encoder = new TextEncoder(); const encodedStrings = strings.map((s) => encoder.encode(s));
  const wordCount = 4 + 1 + encodedStrings.reduce((n, s) => n + 1 + Math.ceil(s.length / 4), 0) + 1 + words.length;
  const buffer = new ArrayBuffer(wordCount * 4); const view = new DataView(buffer); let offset = 0;
  const u32 = (value) => { view.setUint32(offset, value >>> 0, true); offset += 4; };
  u32(BRIDGE_MAGIC); u32(BRIDGE_VERSION); u32(0); u32(ops.length); u32(encodedStrings.length);
  for (const bytes of encodedStrings) { u32(bytes.length); new Uint8Array(buffer, offset, bytes.length).set(bytes); offset += Math.ceil(bytes.length / 4) * 4; }
  u32(words.length); for (const word of words) u32(word);
  return new Uint8Array(buffer, 0, offset);
}

export function decodeBridgeBatch(bytes) {
  const view = new DataView(bytes.buffer, bytes.byteOffset, bytes.byteLength); let offset = 0;
  const u32 = () => { const value = view.getUint32(offset, true); offset += 4; return value; };
  const magic = u32(); const version = u32(); const flags = u32(); const opCount = u32();
  if (magic !== BRIDGE_MAGIC) throw new Error(`Invalid Arrange bridge magic: 0x${magic.toString(16)}`);
  if (version !== BRIDGE_VERSION) throw new Error(`Unsupported Arrange bridge version: ${version}`);
  const decoder = new TextDecoder(); const stringCount = u32(); const strings = [];
  for (let i = 0; i < stringCount; i++) { const length = u32(); strings.push(decoder.decode(new Uint8Array(bytes.buffer, bytes.byteOffset + offset, length))); offset += Math.ceil(length / 4) * 4; }
  const wordCount = u32(); const words = []; for (let i = 0; i < wordCount; i++) words.push(u32());
  let index = 0; const ops = []; const str = (i) => strings[i];
  for (let i = 0; i < opCount; i++) {
    const code = words[index++]; const op = opcodeName[code];
    switch (op) {
      case "createNode": ops.push({ op, id: words[index++], nodeType: str(words[index++]) }); break;
      case "deleteNode": ops.push({ op, id: words[index++] }); break;
      case "insertChild": ops.push({ op, parent: words[index++], child: words[index++], index: words[index++] }); break;
      case "removeChild": ops.push({ op, parent: words[index++], child: words[index++] }); break;
      case "setProp": ops.push({ op, id: words[index++], key: str(words[index++]), value: decodeValue(str(words[index++])) }); break;
      case "setModifier": ops.push({ op, id: words[index++], modifier: decodeValue(str(words[index++])) }); break;
      case "setText": ops.push({ op, id: words[index++], text: str(words[index++]) }); break;
      default: throw new Error(`Unknown Arrange bridge opcode: ${code}`);
    }
  }
  return { header: { magic, version, flags, opCount }, ops };
}
function encodeValue(value) { if (value === undefined) return "u:"; if (value === null) return "n:"; if (typeof value === "number") return `f:${value}`; if (typeof value === "boolean") return `b:${value ? 1 : 0}`; if (typeof value === "string") return `s:${value}`; return `o:${JSON.stringify(value, (_key, val) => typeof val === "function" ? "[Function]" : val)}`; }
function decodeValue(value) { const kind = value.slice(0, 2); const body = value.slice(2); if (kind === "u:") return undefined; if (kind === "n:") return null; if (kind === "f:") return Number(body); if (kind === "b:") return body === "1"; if (kind === "s:") return body; if (kind === "o:") return JSON.parse(body); throw new Error(`Unknown encoded bridge value kind: ${kind}`); }

File: native/arrange_core/src/Bridge.cpp
#include <arrange/core/Bridge.h>

#include <cstring>
#include <limits>

namespace arrange::core {
    namespace {
        class Reader {
        public:
            explicit Reader(std::span<const std::byte> bytes) : bytes_(bytes) {}

            std::uint32_t u32() {
                ensure(4);
                const auto* p = reinterpret_cast<const unsigned char*>(bytes_.data() + offset_);
                const std::uint32_t value = static_cast<std::uint32_t>(p[0]) |
                    (static_cast<std::uint32_t>(p[1]) << 8) |
                    (static_cast<std::uint32_t>(p[2]) << 16) |
                    (static_cast<std::uint32_t>(p[3]) << 24);
                offset_ += 4;
                return value;
            }

            std::string string(std::uint32_t length) {
                ensure(length);
                std::string result(length, '\0');
                if (length > 0) std::memcpy(result.data(), bytes_.data() + offset_, length);
                offset_ += padded(length);
                if (offset_ > bytes_.size()) throw BridgeDecodeError("Bridge string padding exceeds buffer size");
                return result;
            }

        private:
            static std::size_t padded(std::size_t length) noexcept { return ((length + 3u) / 4u) * 4u; }
            void ensure(std::size_t count) const { if (count > bytes_.size() || offset_ > bytes_.size() - count) throw BridgeDecodeError("Unexpected end of bridge command buffer"); }

            std::span<const std::byte> bytes_;
            std::size_t offset_ = 0;
        };

        const std::string& atString(const std::vector<std::string>& strings, std::uint32_t index) {
            if (index >= strings.size()) throw BridgeDecodeError("Bridge string table index out of range");
            return strings[index];
        }

        std::uint32_t atWord(const std::vector<std::uint32_t>& words, std::size_t& index) {
            if (index >= words.size()) throw BridgeDecodeError("Bridge opcode argument stream ended early");
            return words[index++];
        }
    }

    NodeType nodeTypeFromBridgeName(std::string_view name) noexcept {
        if (name == "Box") return NodeType::Box;
        if (name == "Row") return NodeType::Row;
        if (name == "Column") return NodeType::Column;
        if (name == "Spacer") return NodeType::Spacer;
        if (name == "Text") return NodeType::Text;
        if (name == "Input") return NodeType::Input;
        if (name == "Image") return NodeType::Image;
        if (name == "Icon") return NodeType::Icon;
        if (name == "Canvas") return NodeType::Canvas;
        return NodeType::Unknown;
    }

    BridgeBatch decodeBridgeBatch(std::span<const std::byte> bytes) {
        Reader reader(bytes);
        BridgeBatch batch;
        batch.header.magic = reader.u32();
        batch.header.version = reader.u32();
        batch.header.flags = reader.u32();
        batch.header.opCount = reader.u32();

        if (batch.header.magic != BridgeMagic) throw BridgeDecodeError("Invalid Arrange bridge magic");
        if (batch.header.version != BridgeVersion) throw BridgeDecodeError("Unsupported Arrange bridge version");

        const auto stringCount = reader.u32();
        std::vector<std::string> strings;
        strings.reserve(stringCount);
        for (std::uint32_t i = 0; i < stringCount; ++i) { strings.push_back(reader.string(reader.u32())); }

        const auto wordCount = reader.u32();
        std::vector<std::uint32_t> words;
        words.reserve(wordCount);
        for (std::uint32_t i = 0; i < wordCount; ++i) words.push_back(reader.u32());

        std::size_t cursor = 0;
        batch.ops.reserve(batch.header.opCount);
        for (std::uint32_t opIndex = 0; opIndex < batch.header.opCount; ++opIndex) {
            BridgeOp op;
            op.opcode = static_cast<BridgeOpcode>(atWord(words, cursor));
            switch (op.opcode) {
            case BridgeOpcode::CreateNode:
                op.id = atWord(words, cursor);
                op.nodeType = atString(strings, atWord(words, cursor));
                break;
            case BridgeOpcode::DeleteNode:
                op.id = atWord(words, cursor);
                break;
            case BridgeOpcode::InsertChild:
                op.parent = atWord(words, cursor);
                op.child = atWord(words, cursor);
                op.index = atWord(words, cursor);
                break;
            case BridgeOpcode::RemoveChild:
                op.parent = atWord(words, cursor);
                op.child = atWord(words, cursor);
                break;
            case BridgeOpcode::SetProp:
                op.id = atWord(words, cursor);
                op.key = atString(strings, atWord(words, cursor));
                op.value = atString(strings, atWord(words, cursor));
                break;
            case BridgeOpcode::SetModifier:
                op.id = atWord(words, cursor);
                op.modifierPayload = atString(strings, atWord(words, cursor));
                break;
            case BridgeOpcode::SetText:
                op.id = atWord(words, cursor);
                op.text = atString(strings, atWord(words, cursor));
                break;
            default:
                throw BridgeDecodeError("Unknown Arrange bridge opcode");
            }
            batch.ops.push_back(std::move(op));
        }

        if (cursor != words.size()) throw BridgeDecodeError("Bridge command buffer contains trailing opcode words");
        return batch;
    }
} // namespace arrange::core
